libbpf-rs/src/query.rs

```diff
@@ -0,0 +1,249 @@
+//! Query the host about BPF
+//!
+//! For example, to list the name of every bpf program running on the system:
+//! ```
+//! use libbpf_rs::query::ProgInfoIter;
+//!
+//! let mut iter = ProgInfoIter::default();
+//! for prog in iter {
+//!     println!("{}", prog.name);
+//! }
+//! ```
+
+use core::ffi::c_void;
+use std::convert::TryFrom;
+use std::mem::size_of;
+use std::string::String;
+use std::time::Duration;
+
+use nix::{errno, unistd::close};
+
+use crate::*;
+
+macro_rules! gen_info_impl {
+    // This magic here allows us to embed doc comments into macro expansions
+    ($(#[$attr:meta])*
+     $name:ident, $info_ty:ty, $uapi_info_ty:ty, $next_id:expr, $fd_by_id:expr) => {
+        $(#[$attr])*
+        #[derive(Default)]
+        pub struct $name {
+            cur_id: u32,
+        }
+
+        impl Iterator for $name {
+            type Item = $info_ty;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                if unsafe { $next_id(self.cur_id, &mut self.cur_id) } != 0 {
+                    return None;
+                }
+
+                let fd = unsafe { $fd_by_id(self.cur_id) };
+                if fd < 0 && errno::errno() == errno::Errno::ENOENT as i32 {
```

Comment:
I don't think this is the right algorithm. Just because this ID is no more, doesn't mean there are no more IDs to iterate over. See https://github.com/torvalds/linux/blob/master/tools/bpf/bpftool/link.c#L273-L294 for how bpftool iterates, I think that one is correct.

Reply:
Whoops, yeah. Copied it wrong